File: Leetcode/tree/538_Convert BST to Greater Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode* root, int& curMax){
        if(root==nullptr) return;
        dfs(root->right, curMax);
        root->val+=curMax; curMax=root->val;
        dfs(root->left, curMax);
    }
public:
    TreeNode* convertBST(TreeNode* root) {
        int curMax = 0;
        dfs(root, curMax);
        return root;
    }
};

File: Leetcode/tree/538_Convert BST to Greater Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        curMax = [0];
        self.dfs(root, curMax);
        return root;
    def dfs(self, root, curMax):
        if not root: return None
        self.dfs(root.right, curMax)
        root.val+=curMax[0]
        curMax[0]=root.val
        self.dfs(root.left, curMax)

